### ISSUE TITLE
ZoneDelegate is underdocumented

### ISSUE THREAD

**Issue by Hixie** [dart-lang/sdk#26455]
ZoneDelegate says:
> When forwarding to parent zones one can't just invoke the parent zone's exposed functions (like Zone.run), but one needs to provide more information (like the zone the run was initiated). Zone callbacks thus receive more information including this ZoneDelegate class. When delegating to the parent zone one should go through the given instance instead of directly invoking the parent zone.

...but that just raises more questions. _Why_ do you need to provide more information? What is "the zone the run was initiated"?

Further, the individual functions on this class have no documentation at all. If you're in a `handleUncaughtHandler` and you want to `run` something in the context of the parent zone, what are the arguments you pass to `parent.run`, for example?

**Comment by lrhn** [comment]
It should probably be documented more.

The tl;dr of it is that Zones are hierarchical and act like instances of classes extending other classes, but are actually dynamically created using implementation functions with no knowledge of the structure..

So we are really simulating something equivalent to class based object inheritance - when you do `myZone.registerCallback(f)`, you can think of that call as being on an object that has inherited an implementation of `registerCallback` from a superclass, and the implementation may call its super implementation as well. 
Except that it's not classes and there is no `this` or `super` available. That's what the different types represent.

The implementation function is 
```dart
void RegisterCallbackHandler(Zone self, ZoneDelegate parent, Zone zone, void function())
```
Here the `self` represents the actual class where the implementation is declared (maybe think of it as the static scope), `parent` represents its `super` link and `zone` is the value of `this`.

The public facing call of `myZone.registerCallback(f)` is transformed into the internal abstraction where `this` becomes is the explicit `zone` parameter and the actual implementation zone becomes `self`.
From there on, you should not leave that abstraction. If you do, say by calling `parentZone.registerCallback(f)` directly instead of using `parentDelegate.registerCallback(zone, f)`, you are breaking the abstraction and losing the correct `this` zone value.


The actual explanation might want to be more concrete and say when you use which interface and why.
